src/main/scala/ai/eto/rikai/sql/model/Registry.scala

```diff
@@ -16,31 +16,115 @@
 
 package ai.eto.rikai.sql.model
 
+import java.net.URI
+
 /**
-  * Model Registry Integration.
+  * Model Registry Integrations.
   */
 trait Registry {
 
   /**
     * Resolve a Model from the specific URI.
     *
     * @param uri is the model registry URI.
+    * @param name is an optional model name. If provided,
+    *             will create the [[Model]] with this name.
+    *
+    * @throws ModelNotFoundException if the model does not exist on the registry.
     *
-    * @return [[Model]] if found, ``None`` otherwise.
+    * @return [[Model]] if found.
     */
-  def resolve(uri: String): Option[Model]
+  @throws[ModelNotFoundException]
+  def resolve(uri: String, name: Option[String] = None): Model
 }
 
 object Registry {
 
-  val MODEL_REGISTRY_IMPL_KEY = "rikai.sql.ml.model_registry.impl"
+  val REGISTRY_IMPL_PREFIX = "rikai.sql.ml.registry."
+  val REGISTRY_IMPL_SUFFIX = ".impl"
+
+  /** Mapping from Model URI schema to the registry. */
+  private var registryMap: Map[String, Registry] = Map.empty
 
-  /** Get a ModelRegistry from its class name. */
-  def get(className: String): Registry = {
-    Class
-      .forName(className)
-      .getDeclaredConstructor()
-      .newInstance()
-      .asInstanceOf[Registry]
+  @throws[ModelResolveException]
+  private def verifySchema(schema: String): Unit = {
+    val schemaRegex = """[a-zA-Z][\w\+]{0,255}"""
+    if (!schema.matches(schemaRegex)) {
+      throw new ModelResolveException(
+        s"Schema '${schema}' does not match '${schemaRegex}'"
+      )
+    }
+  }
+
+  /**
+    * Register all registry implementations.
+    *
+    * @param conf a mapping of (key, value) pairs
+    */
+  def registerAll(conf: Map[String, String]): Unit = {
+    for ((key, value) <- conf) {
```

Comment:
not a blocker but in case you're interested in a slightly more functional approach, something like the following:

```

  @throws[ModelRegistryAlreadyExistException]
  private def verifyUniqueSchema(schema: String): Unit = {
    if (registryMap.contains(schema)) {
      throw new ModelRegistryAlreadyExistException(
        s"ModelRegistry(${schema}) exists"
        )
    }
  }

  /**
    * Register all registry implementations.
    *
    * @param conf a mapping of (key, value) pairs
    */
  def registerAll(conf: Map[String, String]): Unit = {
    val isSchema = (key: String) => {
      key.startsWith(REGISTRY_IMPL_PREFIX) && key.endsWith(REGISTRY_IMPL_SUFFIX)
    }
    val getSchema = (key: String) => {
      key.substring(REGISTRY_IMPL_PREFIX.length, key.length - REGISTRY_IMPL_SUFFIX.length)
    }
    val toRegistry = (name: String) => {
      Class.forName(name)
           .getDeclaredConstructor(classOf[Map[String, String]])
           .newInstance(conf)
           .asInstanceOf[Registry]
    }
    registryMap ++= conf map { case (key, value) if (isSchema(key)) =>
      val schema = getSchema(key)
      verifySchemaExpr(schema)
      verifyUniqueSchema(schema)
      (schema, toRegistry(value))
    }
  }
```

Reply:
TIL

Reply:
This results

``` scala.MatchError: (spark.driver.host,192.168.86.75) (of class scala.Tuple2)
[info]   at ai.eto.rikai.sql.model.Registry$.$anonfun$registerAll$4(Registry.scala:94)
[info]   at scala.collection.TraversableLike.$anonfun$map$1(TraversableLike.scala:273)
```

seems from ``` registryMap ++= conf map { case (key, value) ```

Reply:
gonna leave it as it is for now

Reply:
needed a slight tweak:

```
registryMap ++= conf.filterKeys(isSchema) map { case (key, value)  =>
      val schema = getSchema(key)
      verifySchemaExpr(schema)
      verifyUniqueSchema(schema)
      (schema, toRegistry(value))
    }
```